### ISSUE TITLE
Question about batch_agent_calls/sync_timesteps/sync_episodes

### ISSUE THREAD

**Issue by joconnor22** [tensorforce/tensorforce#732]
Hi, many thanks again for your efforts on this work.

I'm not quite sure I fully understand the effects of the `batch_agent_calls`, `sync_timesteps` and `sync_episodes` options. I've been playing around with the following MWE to try and understand but still not totally sure so I was hoping you could shed some light.

``` python
from tensorforce import Agent, Environment, Runner
from Helpers import Helper

# Number of environments
nEnvs = 8

# Create environments
environments = []
for _ in range(nEnvs):
    environments.append(dict(environment='Helpers.CustomGym', level='CartPole-v1', seed=0, visualize=False, max_episode_timesteps=500))

# Create agent
agent = Agent.create(
    agent='ppo',
    environment=Environment.create(environments[0]),
    batch_size=32,
    parallel_interactions=nEnvs,
    config=dict(seed=0),
)

# Create and run runner
runner = Runner(agent=agent, environments=environments, num_parallel=nEnvs, remote='multiprocessing')
runner.run(num_episodes=128, callback=Helper.callback, batch_agent_calls=False, sync_timesteps=False, sync_episodes=False, use_tqdm=False)

# Close
runner.close()
agent.close()
```

Note that the custom environment is only used to set the seed (for deterministic learning).

Depending on which combination of options I use there are four different learning outcomes:

1. Everything set to False or just `sync_timesteps` set to True
2. Just `batch_agent_calls` set to True or both `batch_agent_calls` and `sync_timesteps` set to True
3. Just `sync_episodes` set to True or both `sync_episodes` and `sync_timesteps` set to True
4. Everything set to True or both `batch_agent_calls` and `sync_episodes` set to True

In other words, the two combinations for each of the four items in the above list results in the same learning.

I'm guessing the reason that `sync_timesteps` has no effect here is because PPO is episode-based. However, I'm not totally sure what's going on with the other options. I'm also confused about why when everything is set to False and there is no synchronisation that the learning is still deterministic.

Also, when `sync_episodes` is set to True the learning procedure has this strange sawtooth pattern with a frequency that matches the number of parallel environments.

![Fig](https://user-images.githubusercontent.com/8613236/93365233-7227e000-f841-11ea-88df-80c0e50bb340.png)

I'm sure these are obvious questions so apologies in advance!

**Comment by AlexKuhnle** [comment]
Hi,
sorry for the delay in responding. Here some comments (which should probably also be in the docs, if they aren't already):

- (2): `batch_agent_calls` currently implies `sync_timesteps`, so `act` and `observe` calls are synced across *all* environments.
- (1): As you say, PPO is episode-based, so `sync_timesteps` is unlikely to show up. What I think would need to happen is that some environments consistently lag behind here and there by one timestep, and eventually other environments take over by an entire episode, and this eventually interferes with the update frequency. Maybe this will show up if you run for (many) more episodes, but unlikely to happen within the length of 140 episodes.
- (3) and (4): Again, as you say, since PPO is episode-based, `sync_episodes` on or off does matter, but additionally `sync_timesteps` doesn't change anything relevant in when updates happen.
- (1)+(2) vs (3)+(4): Since episode lengths vary quite a bit, `sync_episodes` will change the order episodes come in and hence what the update batches look like.
- (1) vs (2) and (3) vs (4): More subtle, but by batching (terminal) observe calls instead of them coming in as they get finished, the order of episodes gets affected as well.

Will come back to the sawtooth pattern...

**Comment by joconnor22** [comment]
Hi Alex

Thanks for your reply. That pretty much confirms what my initial assumptions were. I think what I hadn't considered is that `sync_timesteps` doesn't imply `sync_episodes` so if one environment finishes an episode early it will sync the first time step of the next episode with the other environments (instead of waiting for all of the other environments to finish). Is that right?

Regarding your second point, that's right. Shortly after opening this issue I tested a different environment with longer episodes and more parallel environments and found the `sync_timesteps` does have an effect.

Final question is does the `blocking` option imply `batch_agent_calls`?

**Comment by AlexKuhnle** [comment]
> so if one environment finishes an episode early it will sync the first time step of the next episode with the other environments (instead of waiting for all of the other environments to finish). Is that right?

Yes, exactly.

> Final question is does the blocking option imply batch_agent_calls?

No. I think it ends up more or less equivalent to `sync_timesteps`. Generally, note that some combinations of these arguments really don't make particular sense, which is probably the case here. In fact, I'm not sure whether there is any real benefit in using `blocking`, but since we don't know what the circumstances of an application may be, no assumptions are made here.

Regarding the sawtooth pattern: first, I thought that this is due to unfinished episodes which have started before an update,  but finish after. However, this doesn't really make sense, but I mention it anyway, since that's a consideration to have in mind (although I think that as long as batch_size/update_frequency and number of environments are aligned accordingly, that shouldn't be a big problem).

My second guess is that this is actually just the distribution of episode lengths within a batch of parallel environments. So within the 8 parallel runs in your case, we always have some longer and some shorter episodes, and since episodes are synced, the shorter ones come in first, the longer ones later. That makes more sense given the 8-episode frequency of spikes.

**Comment by joconnor22** [comment]
Hi Alex

Ah yes that makes sense. Thanks a lot for your help that's cleared everything up for me. I'll close the issue now.
